chain/lf3/ec.go

```diff
@@ -219,3 +205,44 @@ func (ec *ecWrapper) getPowerTableLotusTSK(ctx context.Context, tsk types.TipSet
 	sort.Sort(powerEntries)
 	return powerEntries, nil
 }
+
+func (ec *ecWrapper) Finalize(ctx context.Context, key gpbft.TipSetKey) error {
+	tsk, err := toLotusTipSetKey(key)
+	if err != nil {
+		return err
+	}
+	if err = ec.Syncer.SyncCheckpoint(ctx, tsk); err != nil {
+		return xerrors.Errorf("checkpointing finalized tipset: %w", err)
+	}
+	return nil
+}
+
+func (ec *ecWrapper) toLotusTipSet(ctx context.Context, ts ec.TipSet) (*types.TipSet, error) {
+	switch tst := ts.(type) {
+	case *f3TipSet:
+		return tst.TipSet, nil
+	default:
+		// Fall back on getting the tipset by key. This path is executed only in testing.
+		return ec.getTipSetFromF3TSK(ctx, ts.Key())
+	}
+}
+
+func (ec *ecWrapper) getTipSetFromF3TSK(ctx context.Context, key gpbft.TipSetKey) (*types.TipSet, error) {
+	tsk, err := toLotusTipSetKey(key)
+	if err != nil {
+		return nil, err
+	}
+	ts, err := ec.ChainStore.GetTipSetFromKey(ctx, tsk)
+	if err != nil {
+		return nil, xerrors.Errorf("getting tipset from key: %w", err)
+	}
+	return ts, nil
+}
+
+func toLotusTipSetKey(key gpbft.TipSetKey) (types.TipSetKey, error) {
+	tsk, err := types.TipSetKeyFromBytes(key)
+	if err != nil {
+		return types.TipSetKey{}, xerrors.Errorf("decoding tpiset key: %w", err)
```

Comment:
Does `TipSetKeyFromBytes` not annotate the error?

Reply:
It passes up the cid.CidFromBytes error which may be too generic for debugging purposes. Happy to remove this if you think it is superfluous.

Reply:
Maybe modify `TipSetKeyFromBytes`?